kvdb/postgres/readwrite_tx.go

```
@@ -0,0 +1,296 @@
+`
+}
+
+// CreateTopLevelBucket creates the top level bucket for a key if it
+// does not exist.  The newly-created bucket it returned.
+func (tx *readWriteTx) CreateTopLevelBucket(key []byte) (walletdb.ReadWriteBucket, error) {
+	if len(key) == 0 {
+		return nil, walletdb.ErrBucketNameRequired
+	}
+
+	table := tx.toTableName(key)
+
+	// If the key can't be mapped to a table name, open the bucket from the
+	// hex table.
+	if table == "" {
+		bucket := newReadWriteBucket(tx, tx.db.hexTableName, nil)
+		return bucket.CreateBucketIfNotExists(key)
+	}
+
+	createTableSql := getCreateTableSql(table)
+	_, err := tx.tx.ExecContext(tx.db.ctx, createTableSql)
+	if err != nil {
+		return nil, err
+	}
+
+	return newReadWriteBucket(tx, table, nil), nil
+}
+
+// DeleteTopLevelBucket deletes the top level bucket for a key.  This
+// errors if the bucket can not be found or the key keys a single value
+// instead of a bucket.
+func (tx *readWriteTx) DeleteTopLevelBucket(key []byte) error {
+	bucket := tx.ReadWriteBucket(key)
+	if bucket == nil {
+		return walletdb.ErrBucketNotFound
+	}
+
+	table := bucket.(*readWriteBucket).table
+
+	_, err := tx.tx.ExecContext(tx.db.ctx, "DROP TABLE public."+table+";")
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.tx.ExecContext(tx.db.ctx, "DELETE FROM "+tx.db.sequenceTableName+" WHERE table_name=$1", table)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Commit commits the transaction if not already committed.
+func (tx *readWriteTx) Commit() error {
+	// Commit will fail if the transaction is already committed.
+	if !tx.active {
+		return walletdb.ErrTxClosed
+	}
+
+	// Try committing the transaction.
+	err := tx.tx.Commit()
+	if err == nil && tx.onCommit != nil {
+		tx.onCommit()
+	}
+
+	// Unlock the transaction regardless of the error result.
+	tx.active = false
+	if tx.readOnly {
+		tx.db.lock.RUnlock()
+	} else {
+		tx.db.lock.Unlock()
+	}
+
+	return nil
```

`return err`

---

This seems to have been forgotten. Or why don't we want to return the error here?